pkg/network/route_cache.go

```diff
@@ -0,0 +1,144 @@
+// +build linux
+
+package network
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/golang/groupcache/lru"
+	"github.com/vishvananda/netlink"
+)
+
+type routeKey struct {
+	source, dest util.Address
+	netns        uint32
+	connFamily   ConnectionFamily
+}
+
+// Route stores info for a route table entry
+type Route struct {
+	Gw      util.Address
+	IfIndex int
+}
+
+type routeTTL struct {
+	eta   int64
+	entry Route
+}
+
+type routeCache struct {
+	sync.Mutex
+	cache  *lru.Cache
+	router Router
+	ttl    time.Duration
+}
+
+const defaultTTL = 2 * time.Minute
+
+// RouteCache is the interface to a cache that stores routes for a given (source, destination, net ns) tuple
+type RouteCache interface {
+	Get(source, dest util.Address, netns uint32) (Route, bool)
+}
+
+// Router is an interface to get a route for a (source, destination, net ns) tuple
+type Router interface {
+	Route(source, dest util.Address, netns uint32) (Route, bool)
+}
+
+// NewRouteCache creates a new RouteCache
+func NewRouteCache(size int, router Router) RouteCache {
+	return newRouteCache(size, router, defaultTTL)
+}
+
+// newRouteCache is a private method used primarily for testing
+func newRouteCache(size int, router Router, ttl time.Duration) *routeCache {
+	return &routeCache{
+		cache:  lru.New(size),
+		router: router,
+		ttl:    ttl,
+	}
+}
+
+func (c *routeCache) Get(source, dest util.Address, netns uint32) (Route, bool) {
+	c.Lock()
+	defer c.Unlock()
+
+	k := newRouteKey(source, dest, netns)
+	entry, ok := c.cache.Get(k)
+	if ok && time.Now().Unix() >= entry.(*routeTTL).eta {
+		c.cache.Remove(k)
+		ok = false
+	}
+
+	if !ok && c.router != nil {
```

Comment:
Is there a point to having the cache if `router` is `nil`? Isn't the only way entries are added, is if `router` is not `nil`?